src/bms/parse/validity.rs

```diff
@@ -0,0 +1,614 @@
+//! Validity checks for BMS data after parsing and/or manual edits.
+//!
+//! This module provides a set of structural validations that are independent of
+//! the parsing process. It can be used after editing `Bms` in-memory to ensure
+//! referential integrity and basic invariants required for correct playback.
+
+use thiserror::Error;
+
+use crate::bms::command::{
+    ObjId,
+    channel::{Key, NoteKind, PlayerSide},
+    time::ObjTime,
+};
+
+use super::model::{Bms, obj::Obj};
+
+/// Empty-related validity entries.
+#[non_exhaustive]
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Error)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub enum ValidityEmpty {
+    /// Text content is empty at the time.
+    #[error("Empty text content at {0:?}")]
+    EmptyTextAt(ObjTime),
+    /// Option content is empty at the time.
+    #[cfg(feature = "minor-command")]
+    #[error("Empty option content at {0:?}")]
+    EmptyOptionAt(ObjTime),
+}
+
+/// Missing-related validity entries.
+#[non_exhaustive]
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Error)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub enum ValidityMissing {
+    /// A note references an ObjId without a corresponding WAV definition.
+    #[error("Missing WAV definition for note object id: {0:?}")]
+    MissingWavForNote(ObjId),
+    /// A BGM references an ObjId without a corresponding WAV definition.
+    #[error("Missing WAV definition for BGM object id: {0:?}")]
+    MissingWavForBgm(ObjId),
+    /// A BGA change references an ObjId without a corresponding BMP/EXBMP definition.
+    #[error("Missing BMP definition for BGA object id: {0:?}")]
+    MissingBmpForBga(ObjId),
```

Comment:
Please remove the prefix named its enum type like this:
```suggestion
    #[error("Missing WAV definition for note object id: {0:?}")]
    WavForNote(ObjId),
    /// A BGM references an ObjId without a corresponding WAV definition.
    #[error("Missing WAV definition for BGM object id: {0:?}")]
    WavForBgm(ObjId),
    /// A BGA change references an ObjId without a corresponding BMP/EXBMP definition.
    #[error("Missing BMP definition for BGA object id: {0:?}")]
    BmpForBga(ObjId),
```